modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1282 @@
+  :rtype: `int(32)`
+*/
+proc tcpConn.socketFd throws {
+  var tempfd:c_int;
+  var err:syserr = ENOERR;
+  on this.home {
+    err = qio_get_fd(this._file_internal, tempfd);
+  }
+  if err then try ioerror(err, "in tcpConn.socketFd");
+  return tempfd;
+}
+
+/*
+  Returns the address of remote socket connection
+  :return: Returns remote address.
+  :rtype: `ipAddr`
+*/
+proc tcpConn.addr throws {
+  return getpeername(this.socketFd);
+}
+
+inline operator !=(in lhs: tcpConn,in rhs: tcpConn) {
+  return lhs.socketFd != rhs.socketFd;
+}
+
+proc tcpConn.writeThis(f) throws {
+  f.write("(","addr:",this.addr,",fd:",this.socketFd,")");
+}
+
+pragma "no doc"
+private extern proc sizeof(e): size_t;
+
+extern "struct event_base" record event_base {};
+extern "struct event" record event {};
+
+extern const EV_TIMEOUT:c_short;
+extern const EV_READ:c_short;
+extern const EV_WRITE:c_short;
+extern const EV_SIGNAL:c_short;
+extern const EV_PERSIST:c_short;
+extern const EV_ET:c_short;
+extern const EV_FINALIZE:c_short;
+extern const EV_CLOSED:c_short;
+extern const EVLOOP_NO_EXIT_ON_EMPTY:c_int;
+
+private extern proc event_base_new():c_ptr(event_base);
+private extern proc event_base_dispatch(base: c_ptr(event_base)):c_int;
+private extern proc event_base_loop(base: c_ptr(event_base), flags: c_int):c_int;
+private extern proc event_base_free(base: c_ptr(event_base));
+private extern proc event_base_got_break(base: c_ptr(event_base)):c_int;
+private extern proc event_base_loopbreak(base: c_ptr(event_base)):c_int;
+private extern proc event_new(base: c_ptr(event_base), fd: c_int, events: c_short, callback: c_fn_ptr, callback_arg: c_void_ptr): c_ptr(event);
+private extern proc event_add(ev: c_ptr(event), timeout: c_ptr(timeval)):c_int;
+private extern proc event_del(ev: c_ptr(event)):c_int;
+private extern proc event_free(ev: c_ptr(event)):c_int;
+private extern proc event_remove_timer(ev: c_ptr(event)):c_int;
+private extern proc evutil_make_socket_nonblocking(fd: c_int):c_int;
+private extern proc libevent_global_shutdown();
+
+extern type pthread_t = c_ulong;
+extern type pthread_attr_t;
+private extern proc pthread_create(thread: c_ptr(pthread_t), const attr: c_ptr(pthread_attr_t), start_routine: c_fn_ptr, arg: c_void_ptr): c_int;
+private extern proc pthread_join(thread: pthread_t, retval: c_ptr(c_void_ptr)): c_int;
+private extern proc pthread_exit(retval: c_void_ptr);
+private extern proc evthread_use_pthreads();
+
+evthread_use_pthreads();
+var event_loop_base = event_base_new();
+
+proc dispatchLoop():c_void_ptr throws {
+  if event_loop_base == nil {
+    throw new Error("event loop wasn't initialized");
+  }
+  var ret_val = 0;
+  if event_base_got_break(event_loop_base) == 1 {
+    pthread_exit(c_ptrTo(ret_val):c_void_ptr);
+  }
+  var x = event_base_loop(event_loop_base, EVLOOP_NO_EXIT_ON_EMPTY);
+  if x != 0 {
+    throw new Error("event loop wasn't initialized");
+  }
+  pthread_exit(c_ptrTo(ret_val):c_void_ptr);
```

Comment:
I would expect that simply returning is sufficient here, rather than calling `pthread_exit`. Do you know of a reason that is not the case?

Note though that if you want to return an integer, you have to copy it in to a `c_void_ptr`, since you can't return a pointer to a local variable without causing memory errors. So I'd recommend writing it something like this:

``` chapel
var retInt : c_intptr = 0;
var retAsPtr : c_void_ptr;
c_memcpy(retAsPtr, retInt, sizeof(retAsPtr));
return retAsPtr;
```

You can use a similar `c_memcpy` approach to get the return value out, but it appears today that you do not use it. If you know you don't need a return value from the `dispatchLoop` here, let's just have it return `nil` instead of doing `pthread_exit` and remove the `ret_val` variable entirely.

Reply:
Will go with `nil` approach